File: Number-Guessing-Game/game.js
#!/usr/bin/env node
/**
 * Number‑Guessing Game – pure Node.js, no external dependencies
 *
 * Features:
 *   • Three difficulty levels (Easy/Medium/Hard) → 10/5/3 guesses
 *   • Random number between 1 and 100
 *   • Hint system (parity + narrowed range) after a few wrong attempts
 *   • Timer (how long the round lasted)
 *   • Persistent high‑score table (fewest attempts per difficulty)
 *   • Play multiple rounds until the user quits
 *
 * Run with:   node game.js
 */

const fs   = require('fs');
const path = require('path');
const readline = require('readline');

// ----------------------------------------------------------------------
// Configuration constants (tweak if you like)
// ----------------------------------------------------------------------
const MIN_NUMBER   = 1;
const MAX_NUMBER   = 100;

const DIFFICULTIES = {
    1: { name: 'Easy',   chances: 10 },
    2: { name: 'Medium', chances: 5  },
    3: { name: 'Hard',   chances: 3  },
};

const HINT_AFTER_WRONG = 3;               // after how many wrong guesses a hint is offered
const HIGHSCORE_FILE   = path.join(__dirname, 'highscores.json');

// ----------------------------------------------------------------------
// Helper: async question prompt (wraps readline.question in a Promise)
// ----------------------------------------------------------------------
function ask(question) {
    const rl = readline.createInterface({
        input:  process.stdin,
        output: process.stdout,
    });
    return new Promise(resolve => rl.question(question, ans => {
        rl.close();
        resolve(ans.trim());
    }));
}

// ----------------------------------------------------------------------
// Persistent high‑score handling (read/write JSON)
// ----------------------------------------------------------------------
function loadHighScores() {
    if (!fs.existsSync(HIGHSCORE_FILE)) return {};
    try {
        const raw = fs.readFileSync(HIGHSCORE_FILE, 'utf8');
        return JSON.parse(raw);
    } catch (_) {
        // Corrupted file – start fresh
        return {};
    }
}

function saveHighScores(scores) {
    try {
        fs.writeFileSync(HIGHSCORE_FILE, JSON.stringify(scores, null, 2), 'utf8');
    } catch (e) {
        console.warn('⚠️  Could not write high‑score file:', e.message);
    }
}

// ----------------------------------------------------------------------
// Utility: read an integer with optional bounds, keep asking until valid
// ----------------------------------------------------------------------
async function readInt(prompt, low = null, high = null) {
    while (true) {
        const ans = await ask(prompt);
        const num = Number(ans);
        if (Number.isInteger(num) && (low === null || num >= low) && (high === null || num <= high)) {
            return num;
        }
        console.log(`Please enter an integer${low!==null?` ≥ ${low}`:''}${high!==null?` ≤ ${high}`:''}.`);
    }
}

// ----------------------------------------------------------------------
// Utility: read a yes/no answer (y/n)
// ----------------------------------------------------------------------
async function readYesNo(prompt) {
    while (true) {
        const ans = (await ask(prompt)).toLowerCase();
        if (ans === 'y' || ans === 'yes') return true;
        if (ans === 'n' || ans === 'no')  return false;
        console.log('Please answer with y (yes) or n (no).');
    }
}

// ----------------------------------------------------------------------
// Hint generator – parity + current narrowed range
// ----------------------------------------------------------------------
function giveHint(target, lowBound, highBound) {
    const parity = target % 2 === 0 ? 'even' : 'odd';
    console.log(`💡  Hint: The number is ${parity} and lies between ${lowBound} and ${highBound}.`);
}

// ----------------------------------------------------------------------
// One round of the game
// ----------------------------------------------------------------------
async function playRound(highScores) {
    console.log('\n=== Welcome to the Number Guessing Game! ===');
    console.log(`I'm thinking of a number between ${MIN_NUMBER} and ${MAX_NUMBER}.\n`);

    // ----------------  choose difficulty -----------------
    console.log('Please select the difficulty level:');
    for (const [key, cfg] of Object.entries(DIFFICULTIES)) {
        console.log(`${key}. ${cfg.name} (${cfg.chances} chances)`);
    }
    const diffChoice = await readInt('\nEnter your choice (1/2/3): ', 1, 3);
    const { name: diffName, chances } = DIFFICULTIES[diffChoice];
    console.log(`\nGreat! You have selected the ${diffName} difficulty level.`);
    console.log("Let's start the game!\n");

    // ----------------  initialise round -----------------
    const secretNumber = Math.floor(Math.random() * (MAX_NUMBER - MIN_NUMBER + 1)) + MIN_NUMBER;
    let attempts = 0;
    let lowerPossible = MIN_NUMBER;
    let upperPossible = MAX_NUMBER;
    const startTime = Date.now();

    // ----------------  main guessing loop ---------------
    while (attempts < chances) {
        const remaining = chances - attempts;
        const guess = await readInt(`Attempt ${attempts + 1}/${chances} – your guess: `, MIN_NUMBER, MAX_NUMBER);
        attempts++;

        if (guess === secretNumber) {
            const elapsedSec = ((Date.now() - startTime) / 1000).toFixed(1);
            console.log(`\n✅  Congratulations! You guessed it in ${attempts} attempt(s).`);
            console.log(`⏱️  Time taken: ${elapsedSec} seconds.`);

            // ---- update high‑score -------------------------------------------------
            const best = highScores[diffName];
            if (best == null || attempts < best) {
                highScores[diffName] = attempts;
                console.log(`🎉  New high score for ${diffName}! (${attempts} attempts)`);
            } else {
                console.log(`Current high score for ${diffName}: ${best} attempts`);
            }
            break;
        }

        // ---- wrong guess ---------------------------------------------------------
        if (guess < secretNumber) {
            console.log('Incorrect! The secret number is greater than your guess.');
            lowerPossible = Math.max(lowerPossible, guess + 1);
        } else {
            console.log('Incorrect! The secret number is less than your guess.');
            upperPossible = Math.min(upperPossible, guess - 1);
        }

        // ---- hint opportunity ----------------------------------------------------
        if (attempts >= HINT_AFTER_WRONG) {
            const wantHint = await readYesNo('Would you like a hint? (y/n): ');
            if (wantHint) giveHint(secretNumber, lowerPossible, upperPossible);
        }

        if (attempts < chances) {
            console.log(`You have ${remaining - 1} chance(s) left.\n`);
        } else {
            console.log('\n❌  You have used all your chances!');
            console.log(`The number was: ${secretNumber}`);
        }
    }

    // ---------- display high‑scores after the round ----------
    console.log('\n=== High Scores (fewest attempts) ===');
    for (const level of ['Easy', 'Medium', 'Hard']) {
        console.log(`${level}: ${highScores[level] ?? '—'}`);
    }
    console.log('=====================================\n');

    return highScores;
}

// ----------------------------------------------------------------------
// Main loop – keep playing until the player quits
// ----------------------------------------------------------------------
async function main() {
    let highScores = loadHighScores();

    while (true) {
        highScores = await playRound(highScores);
        saveHighScores(highScores);

        const again = await readYesNo('Do you want to play again? (y/n): ');
        if (!again) {
            console.log('\nThanks for playing! Goodbye 👋');
            break;
        }
    }
}

// ----------------------------------------------------------------------
// Run the game, handling Ctrl‑C gracefully
// ----------------------------------------------------------------------
main().catch(err => {
    console.error('Unexpected error:', err);
    process.exit(1);
});

File: Number-Guessing-Game/game.py
#!/usr/bin/env python3
"""
Number Guessing Game – CLI version

Features
--------
* Three difficulty levels (Easy / Medium / Hard)
* Limited number of guesses per level
* Hint system (parity + narrowed range) after a few wrong attempts
* Timer that shows how long the round lasted
* High‑score table (fewest attempts per difficulty) saved to disk
* Play multiple rounds until the user decides to quit

Run with:
    python game.py
"""

import json
import os
import random
import sys
import time
from typing import Optional  # <-- needed for the corrected type hints

# ----------------------------------------------------------------------
# Configuration constants (feel free to tweak)
# ----------------------------------------------------------------------
MIN_NUMBER = 1
MAX_NUMBER = 100

DIFFICULTIES = {
    "1": {"name": "Easy",   "chances": 10},
    "2": {"name": "Medium", "chances": 5},
    "3": {"name": "Hard",   "chances": 3},
}

HINT_AFTER_WRONG = 3               # after how many wrong guesses a hint is offered
HIGHSCORE_FILE = "highscores.json"  # persisted high‑score file (same folder)

# ----------------------------------------------------------------------
# Small I/O helpers
# ----------------------------------------------------------------------
def clear_line():
    """Print an empty line – helps readability."""
    print()


def prompt_int(
    prompt: str,
    low: Optional[int] = None,
    high: Optional[int] = None,
) -> int:
    """
    Read an integer from stdin, optionally enforce low/high bounds.

    Returns the integer the user entered.
    """
    while True:
        val = input(prompt).strip()
        if val.isdigit() or (val.startswith("-") and val[1:].isdigit()):
            num = int(val)
            if (low is not None and num < low) or (high is not None and num > high):
                print(f"Please enter a number between {low} and {high}.")
                continue
            return num
        else:
            print("That does not look like a number – try again.")


def prompt_choice(prompt: str, choices: list) -> str:
    """Read a choice from a list of allowed strings (case‑insensitive)."""
    choices_lower = [c.lower() for c in choices]
    while True:
        val = input(prompt).strip().lower()
        if val in choices_lower:
            return val
        else:
            print(f"Please type one of: {', '.join(choices)}")


# ----------------------------------------------------------------------
# High‑score handling
# ----------------------------------------------------------------------
def load_highscores() -> dict:
    """Return a dict like {'Easy': 7, 'Medium': 4, 'Hard': 3}."""
    if not os.path.exists(HIGHSCORE_FILE):
        return {}
    try:
        with open(HIGHSCORE_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        # Corrupted file – start fresh
        return {}


def save_highscores(highscores: dict):
    """Write the high‑score dict to disk (pretty‑printed)."""
    try:
        with open(HIGHSCORE_FILE, "w", encoding="utf-8") as f:
            json.dump(highscores, f, indent=2, sort_keys=True)
    except OSError as e:
        print(f"⚠️  Could not save high scores: {e}")


def maybe_update_highscore(highscores: dict, difficulty_name: str, attempts: int):
    """Update the dict if the new result beats the stored record."""
    record = highscores.get(difficulty_name)
    if record is None or attempts < record:
        highscores[difficulty_name] = attempts
        print(f"🎉  New high score for {difficulty_name} difficulty: {attempts} attempts!")
    else:
        print(f"Current high score for {difficulty_name}: {record} attempts.")
    return highscores


# ----------------------------------------------------------------------
# Hint system
# ----------------------------------------------------------------------
def give_hint(target: int, low_bound: int, high_bound: int):
    """
    Print a simple hint:
    * Whether the number is even or odd.
    * The current narrowed range (low_bound … high_bound).
    """
    parity = "even" if target % 2 == 0 else "odd"
    print(f"💡  Hint: The number is {parity} and lies between {low_bound} and {high_bound}.")


# ----------------------------------------------------------------------
# One round of the game
# ----------------------------------------------------------------------
def play_round(highscores: dict) -> dict:
    """Run a single guessing round and return the (potentially updated) high‑score dict."""
    print("\n=== Welcome to the Number Guessing Game! ===")
    print(f"I'm thinking of a number between {MIN_NUMBER} and {MAX_NUMBER}.\n")

    # ------------------------------------------------------------------
    # Choose difficulty
    # ------------------------------------------------------------------
    print("Please select the difficulty level:")
    for key, data in DIFFICULTIES.items():
        print(f"{key}. {data['name']} ({data['chances']} chances)")
    difficulty_key = prompt_choice("\nEnter your choice (1/2/3): ", list(DIFFICULTIES.keys()))
    difficulty = DIFFICULTIES[difficulty_key]
    chances = difficulty["chances"]
    diff_name = difficulty["name"]
    clear_line()
    print(f"Great! You have selected the {diff_name} difficulty level.")
    print("Let's start the game!\n")

    # ------------------------------------------------------------------
    # Initialise game state
    # ------------------------------------------------------------------
    target_number = random.randint(MIN_NUMBER, MAX_NUMBER)
    attempts = 0
    lower_possible = MIN_NUMBER
    upper_possible = MAX_NUMBER

    start_time = time.time()

    while attempts < chances:
        remaining = chances - attempts
        guess = prompt_int(
            f"Enter your guess (attempt {attempts + 1}/{chances}): ",
            low=MIN_NUMBER,
            high=MAX_NUMBER,
        )
        attempts += 1

        # --------------------------------------------------------------
        # Correct guess?
        # --------------------------------------------------------------
        if guess == target_number:
            elapsed = time.time() - start_time
            minutes = int(elapsed // 60)
            seconds = elapsed % 60
            print(f"\n✅  Congratulations! You guessed the number in {attempts} attempts.")
            print(f"⏱️  Time taken: {minutes}m {seconds:.1f}s")
            # Update high‑score table
            highscores = maybe_update_highscore(highscores, diff_name, attempts)
            break

        # --------------------------------------------------------------
        # Wrong guess – give feedback & tighten range
        # --------------------------------------------------------------
        if guess < target_number:
            print("Incorrect! The number is greater than your guess.")
            lower_possible = max(lower_possible, guess + 1)
        else:
            print("Incorrect! The number is less than your guess.")
            upper_possible = min(upper_possible, guess - 1)

        # Offer a hint after a few wrong attempts
        if attempts >= HINT_AFTER_WRONG:
            want_hint = prompt_choice("Would you like a hint? (y/n): ", ["y", "n"])
            if want_hint == "y":
                give_hint(target_number, lower_possible, upper_possible)

        if attempts < chances:
            print(f"You have {remaining - 1} chances left.\n")
        else:
            # Ran out of chances – reveal the answer
            print("\n❌  You've run out of chances!")
            print(f"The correct number was: {target_number}")
            print("Better luck next time.\n")

    # ------------------------------------------------------------------
    # End‑of‑round summary
    # ------------------------------------------------------------------
    print("\n=== High Scores ===")
    for level in ["Easy", "Medium", "Hard"]:
        score = highscores.get(level, "—")
        print(f"{level}: {score}")
    print("===================\n")

    return highscores


# ----------------------------------------------------------------------
# Main loop – keep playing until the user quits
# ----------------------------------------------------------------------
def main():
    # Load (or initialise) the persistent high‑score table
    highscores = load_highscores()

    while True:
        highscores = play_round(highscores)
        # Persist after each round
        save_highscores(highscores)

        again = prompt_choice("Do you want to play again? (y/n): ", ["y", "n"])
        if again != "y":
            print("\nThanks for playing! Goodbye 👋")
            break
        clear_line()


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        # Graceful exit on Ctrl‑C
        print("\n\n👋  Game interrupted. Bye!")
        sys.exit(0)
